### PR TITLE
Invalidate cached theme dictionary when theme dictionaries are added/removed

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): Part of #7123

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Cached value doesn't get properly updated when needed.

## What is the new behavior?

Update cached value properly.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_ResourceDictionary.cs [hunk 1]
```diff
@@ -795,5 +795,20 @@ public void When_Custom_Resource_Dictionary_With_Custom_Property_in_Custom_Contr
...
+
+		[TestMethod]
+		public void When_Theme_Dictionary_Is_Cached()
+		{
+			var dictionary = new ResourceDictionary();
+			dictionary.TryGetValue("TestKey", out _); // This causes _activeThemeDictionary to be cached.
+
+			var lightTheme = new ResourceDictionary();
+			lightTheme.Add("TestKey", "TestValue");
+			dictionary.ThemeDictionaries.Add("Light", lightTheme); // Cached value is no longer valid.
```
Comment (Youssef1313) [inline src/Uno.UI.Tests/Windows_UI_Xaml/Given_ResourceDictionary.cs:807]: I'm not sure hard coding "Light" here is correct.

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 2]
```diff
@@ -48,7 +72,33 @@ public Uri Source
...
+				_themeDictionaries = new ResourceDictionary();
+				_themeDictionaries.ResourceDictionaryValueChange += (sender, e) =>
+				{
+					// The purpose of this to updated the cached value when it's no longer valid.
+					// See GetActiveThemeDictionary.
+					if (e.ChangedKey is ResourceKey changedKey && changedKey.Equals(_activeTheme))
+					{
+						// Either a key is added (e.NewValue will be non-null), or
+						// a key is removed (e.NewValue will be null).
+						_activeThemeDictionary = e.NewValue;
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:90]: I wonder what happens if the dictionary was falling back on the "Default" theme (or now should be), though we may be getting into the edge case of an edge case.

I also wonder how the case of nested dictionaries in ThemeDictionaries is handled (both by this and by the existing caching logic), since ResourceDictionary.TryGetValue (which is called by GetActiveThemeDictionary()) will search into nested (merged and theme) dictionaries.

I'm curious what the actual scenario was that you ran into where you noticed this bug?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:90]: > I'm curious what the actual scenario was that you ran into where you noticed this bug?

@davidjohnoliver It was a simple port of NumberBox from microsoft-ui-xaml repo that revealed the bug.

---

> I wonder what happens if the dictionary was falling back on the "Default" theme (or now should be), though we may be getting into the edge case of an edge case.

This is really very interesting. I'll see if I can reproduce in a unit test and look into how would I fix it.

**Comment by Youssef1313** [comment]
@davidjohnoliver Can you see the updated logic please?

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs (hunk 2 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs]: I get that it was related to the earlier NumberBox PR, but I was wondering about what was actually happening that triggered this case.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs]: @davidjohnoliver I couldn't really know the answer. @jeromelaban Do you know?

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 3]
```diff
@@ -48,7 +72,41 @@ public Uri Source
...
+					// The purpose of this to updated the cached value when it's no longer valid.
+					// See GetActiveThemeDictionary.
+					if (e.ChangedKey is ResourceKey changedKey && changedKey.Equals(_activeTheme))
+					{
+						if (e.NewValue is null && !_activeTheme.Equals(Themes.Default))
+						{
+							// If the key is removed and we have non-default active theme,
+							// then update cached value with default theme.
+							// We do that here because GetActiveThemeDictionary does that.
+							_activeThemeDictionary = GetThemeDictionary(Themes.Default);
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: Since you're now setting _activeTheme from GetThemeDictionary(), this will set _activeTheme to Default, which seems strange?

Instead of trying to recalculate _activeThemeDictionary here, can you simply set _activeTheme to ResourceKey.Empty? Then _activeThemeDictionary will be correctly re-resolved the next time `GetActiveThemeDictionary()` is called.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: Doing so is still problematic for the following case:

- Active theme is Light but we only have Default
- We assign `Light` to `_activeTheme`, and the default dictionary to `_activeThemeDictionary` (through the `?? GetThemeDictionary(Themes.Default)`
- Later, "Default" gets removed

For this scenario, we shouldn't have anything available, but we'll be caching the "Default" theme if the `_activeTheme` wasn't set to `Default`. Is there a better approach that handles this scenario?

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: But it's problematic in the case that we have cached the "Default" dictionary, and later on "Light" gets added, or no?

Can you just invalidate `_activeTheme` when _any_ dictionary is added/removed from ThemeDictionaries? That seems safest, and modifying `ThemeDictionaries` is sufficiently uncommon that it hopefully shouldn't pose a performance problem.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: > But it's problematic in the case that we have cached the "Default" dictionary, and later on "Light" gets added, or no?

In that case, the next time `GetActiveThemeDictionary` is called won't use the cached value because `_activeTheme` is "Default" but we're requesting "Light". So it shouldn't be problematic?

Anyway I agree that invalidating _activeTheme whenever a dictionary is added or removed is safest. But I want to make sure we have a good test coverage in case someone adds the caching back in future :)

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: > In that case, the next time GetActiveThemeDictionary is called won't use the cached value because _activeTheme is "Default" but we're requesting "Light". So it shouldn't be problematic?

It sounds like it's the caching itself that would be broken in that case, but I can't see the code anymore, so never mind. :D

And yes, whatever the fix, good test coverage is useful. ;)

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:93]: @davidjohnoliver Do you mean the current implementation is okay, or I should still be invalidating the cache for any theme dictionary change? or both are okay (in which case I'd like to see if @jeromelaban has any preference)?

File: src/Uno.UI.Tests/Windows_UI_Xaml/Given_ResourceDictionary.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.UI.Tests/Windows_UI_Xaml/Given_ResourceDictionary.cs]: You could use a nameof from the RequestedTheme, but that's unlikely to change.

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs (hunk 2 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs]: This is happening because some resources initialized in the NumberBox updates are changing the initialization order, causing some early evaluations. It could be something related to the lazy evaluation you've been looking into @davidjohnoliver.

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 4]
```diff
@@ -21,6 +21,30 @@ public partial class ResourceDictionary : DependencyObject, IDependencyObjectPar
 		private readonly List<ResourceDictionary> _mergedDictionaries = new List<ResourceDictionary>();
 		private ResourceDictionary _themeDictionaries;
 
+		/// <summary>
+		/// This event is fired when a key that has value of type <see cref="ResourceDictionary"/> is added or changed in the current <see cref="ResourceDictionary" />
+		/// </summary>
+		private event EventHandler<ResourceDictionaryChangedEventArgs> ResourceDictionaryValueChange;
+
+		private class ResourceDictionaryChangedEventArgs
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:29]: This class is no longer needed since we invalidate the cache for any changes now. But it may (or may not) be used later. I'm not sure if I should remove it or keep it. @davidjohnoliver @jeromelaban What do you think?

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/ResourceDictionary.cs:29]: Personally I would remove it if it's not used, it's internal so there's no difficulty in bringing it back if it becomes necessary.

**Review by jeromelaban — APPROVED** [review]
Thanks for your contribution!
